### PR TITLE
Support Heterogeneous Lookup

### PR DESCRIPTION
**Pull request by dalloriam**
This PR introduces support for heterogeneous lookup (using a different key type for lookups) in `cachemere::Cache`.

## Public-facing Changes
- Introduces two new primitives (one internal, one public) used to implement heterogeneous lookup:
  - [TransparentEq](https://github.com/coveooss/cachemere/pull/31/files#diff-8f47328eb2fca1d3815b382ba939f87b232762bffe39b55ce079c9555b7dafcaR8), which is a generic equality functor used internally in the cache to check if a given type is equatable to the cache key type.
  - [MultiHash](https://github.com/coveooss/cachemere/pull/31/files#diff-8072929d4e60482b03d303c38aa93fff8022abe28fd9f159b44ae7df0990cebbR11) - which uses dark & forbidden template magic to generate hashing overloads for a set of types. 
    - e.g. `MultiHash<int, absl::Hash<int>, std::string, std::hash<std::String>>`
- Introduces the `KeyView` template parameter on `Cache::find()` and `Cache::contains()`
  - For a `KeyView` to be valid, it needs to implement `bool operator==(const Key&)` and be hashable by `KeyHash` (use of `MultiHash` is recommended)

## Internal Changes
- Removes the `Item` template parameter on `BloomFilter` and `CountingBloomFilter` in order to support heterogeneous lookup in those datastructrues

### CODE REVIEW

File: heterogeneousLookup.dox [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+using CacheT = cachemere::presets::memory::LRUCache<CompositeKey,
+                                                    Value,
+                                                    cachemere::measurement::CapacityDynamicallyAllocated<Value>,
+                                                    cachemere::measurement::CapacityDynamicallyAllocated<CompositeKey>,
+                                                    MultiHasher>;
+
+bool contains(const CacheT& cache, std::string_view a, const std::string_view b, const std::string_view c) {
+    // We can use the lightweight key directly, without allocating a single string.
+    const CompositeView key{std::move(a), std::move(b), std::move(c)};
```
Comment (klalumiere) [inline heterogeneousLookup.dox]: nitpick: It's probable more (or as) efficient to copy `std::string_view` than to `move` them. :nerd_face:

File: heterogeneousLookup.dox [hunk 2]
```diff
@@ -0,0 +1,113 @@
+Using Heterogeneous Lookup {#heterogeneousLookup}
```
Comment (klalumiere) [inline heterogeneousLookup.dox:1]: Thanks for the doc! :heart:

File: include/cachemere/detail/transparent_eq.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#ifndef CACHEMERE_TRANSPARENT_EQ_H
+#define CACHEMERE_TRANSPARENT_EQ_H
+
+#include <absl/hash/hash.h>
+
+namespace cachemere::detail {
+
+template<typename Key> struct TransparentEq {
+    using is_transparent = void;
```
Comment (klalumiere) [inline include/cachemere/detail/transparent_eq.h:10]: nitpick: while reading the code, I stop to ask myself why this line was there. Then, in the next file, I read `// Declare this type as transparent - this is needed for abseil maps to support heterogeneous lookup.`. Maybe you could put the comment there too. Or not, it's not a big deal. :slightly_smiling_face:

File: justfile [hunk 4]
```diff
@@ -0,0 +1,10 @@
+configure cfg="Debug" dir="build":
+    cmake -B {{dir}} -S . -DUSE_VCPKG=ON -DBUILD_TESTS=ON -DBUILD_BENCHMARKS=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DCMAKE_BUILD_TYPE={{cfg}}
+
+build dir="build":
+    cmake --build {{dir}}
+
+test cfg="Debug" dir="build":
+    @just configure {{cfg}} {{dir}}
+    @just build {{dir}}
+    {{dir}}/tests/cachemere_tests
```
Comment (klalumiere) [inline justfile:10]: thought: There's something funny in seing a Make-like program being used to call CMake which is itself a build system generator that (can) generates Make files to build the program. :laughing: It feels like

image

File: tests/src/cache_tests.cpp [hunk 5]
```diff
@@ -377,3 +383,81 @@ TEST(CacheTest, SingleThreadSwapDoesntThrow)
...
+
+    TestCache cache{100 * sizeof(Point3D)};
+
+    CompositeKey     key{"asdf", "hjkl"};
+    CompositeKeyView key_view{"asdf", "hjkl"};
+
+    Point3D value{1, 1, 1};
+
+    cache.find(key_view);
+    cache.insert(key, value);
```
Comment (klalumiere) [inline tests/src/cache_tests.cpp:439]: Why do you call `find` before inserting? Do you want to show that it doesn't throw? Maybe you should add a comment to clarify your intent. :thinking: Or you could verify that the `find` returned "not found".

File: tests/src/cache_tests.cpp [hunk 6]
```diff
@@ -377,3 +383,81 @@ TEST(CacheTest, SingleThreadSwapDoesntThrow)
...
+
+    CompositeKey     key{"asdf", "hjkl"};
+    CompositeKeyView key_view{"asdf", "hjkl"};
+
+    Point3D value{1, 1, 1};
+
+    EXPECT_FALSE(cache.contains(key_view));
+
+    cache.find(key);
+    cache.insert(key, value);
```
Comment (klalumiere) [inline tests/src/cache_tests.cpp]: Why do you call `find` before inserting? Do you want to show that it doesn't throw? Maybe you should add a comment to clarify your intent. :thinking: Or you could verify that the `find` returned "not found". 

Note that in the previous test, you call `cache.find(key_view);` and here, you call `cache.find(key);`. I don't understand why it's different. :thinking:

**Review by klalumiere — APPROVED** [review]
Very nice!

The two comments in `tests/src/cache_tests.cpp` may worth a few seconds to address. The rest is nitpick. :slightly_smiling_face:

File: tests/src/cache_tests.cpp (hunk 6 above)
Comment (dalloriam) [reply tests/src/cache_tests.cpp]: I agree that it's not clear, I'll add a comment but I'll also explain here:

The `TinyLFU` algorithm uses a frequency sketch to compute whether a key should be added into the cache, and for it to be accepted the key needs to have been seen at least once.

**Review by cbendellcoveo — APPROVED** [review]

**Pull request merged**